### ISSUE TITLE
Add support for charge detection and wifi detection in Android

### ISSUE THREAD

**Issue by cereallarceny** [OpenMined/KotlinSyft#37]

**Comment by galalen** [comment]
I want to work on this issue

**Comment by vkkhare** [comment]
Hi @galalen we would like to implement a method that could check for the battery status and be passed as an argument to syftlib. Check this
https://developer.android.com/training/monitoring-device-state/battery-monitoring.html

Since the context object is only available in the app implementing the syftlib, this has to be passed by the demo_app. Let's create a `SyftConfiguration.kt` in syftlib which will take the context object and define the functions `getWifiStatus`. `Syft.kt` can then use these functions for enforcing checks

**Comment by galalen** [comment]
Great, @vkkhare. @cereallarceny I think I'm gonna follow @vkkhare way. 
My proposed solution is not that different I thought of creating a utility class in syftlib which receives the context and has two methods `isWifiConnected `and `isDeviceChargerConnected`.

Now I'm thinking of adding a Broadcast Receiver to notify the user of behalf the android system with the battery state.

**Comment by mccorby** [comment]
I was thinking of having these Android specific functionality in a separate module. This was we can make `syftlib` as non-Android at all.
`syftlib` should just expose a contract (an interface) that is implemented in that other module and provided by the client application.

Eventually we could see if that new module can be exported as another library.

**Comment by galalen** [comment]
Make sense, it can be a utility for syft which contains all core android related staff

**Comment by vkkhare** [comment]
Aahh yes that will help you in porting KotlinSyft for desktop clients as well @mccorby ?

**Comment by mccorby** [comment]
We can enforce it by checking if certain classes exist when starting `Syft`. Or checking a properties file. There are different ways of doing it. I would put this as a separate issue.

Also, we need to think the users of the library are somehow intelligent and responsible beings ;)

**Comment by mccorby** [comment]
Having the charge and wifi detection in a separate library would also force the users to select one of the implementations (for instance, `syft-detection-android` vs `syft-detection-desktop`).

**Comment by cereallarceny** [comment]
Is there any way to just provide some sort of observable/event listener
hook in the config?  If the implementer wants to have WiFi/charge
detection, they can provide an interface where that can be “listened” to by
the library.  However, it’s not a requirement that WiFi and charge
detection exist for training.  That way, the person implementing is in full
control of that functionality, and when it does or doesn’t apply.  All
you’d need to do is just provide the “hook”.  Am I making sense?

**Comment by mccorby** [comment]
Yes, that is the idea: `Syft` expects an implementation of the interface, the client can pass anything complying with its contract.
In training, it can pass a dummy object (something that should be used in tests).
The point is for `Syft` to not have any actual implementation of how the wifi/charge detection works.

**Comment by cereallarceny** [comment]
Perfect.  Go ahead then!

**Comment by vkkhare** [comment]
If we can enforce this, no problems from my side

**Comment by vkkhare** [comment]
Blocked until #87 is finished

**Comment by vkkhare** [comment]
AN alternate path using #89 has been added and it no more blocked by background execution

**Comment by cereallarceny** [comment]
Great, close this if you have a reasonable alternative.
